### ISSUE TITLE
http/2: Large transfers have high overhead in Curl_bufq_len

### ISSUE THREAD

**Issue by laramiel** [curl/curl#13403]
### I did this

In [tensorstore](https://github.com/google/tensorstore) we (currently) use a single multi-handle to multiplex a lot of http/2 transfers.  Those transfers appear to be slower than expected.

The following benchmark, which downloads 1000 files over http/2 (limited to 32 concurrently), where each file is 64MB, demonstrates the issue.

I get output something like this, which is to say, 31% of the time is spent in `Curl_bufq_len`

```
638.04 s	31.91%  Curl_bufq_len (bufq.c)
638.04 s	31.91%  on_stream_frame (http2.c) inlined
638.04 s	31.91%  on_frame_recv (http2.c)
638.04 s	31.91%  session_call_on_frame_received (nghttp2_session.c) inlined
638.04 s	31.91%  nghttp2_session_on_data_received (nghttp2_session.c)
638.04 s	31.91%  session_process_data_frame (nghttp2_session.c) inlined
638.04 s	31.91%  nghttp2_session_mem_recv (nghttp2_session.c)
638.04 s	31.91%  h2_process_pending_input (http2.c)
638.04 s	31.91%  h2_progress_ingress (http2.c)
638.04 s	31.91%  cf_h2_recv (http2.c)
638.04 s	31.91%  Curl_read (sendf.c)
638.04 s	31.91%  Curl_xfer_recv_resp (transfer.c) inlined
638.04 s	31.91%  readwrite_data (transfer.c) inlined
638.04 s	31.91%  Curl_readwrite (transfer.c)
638.04 s	31.91%  multi_runsingle (multi.c)
```

Repro:

```
git clone https://github.com/google/tensorstore.git

./bazelisk.py build -c opt --copt=-g tensorstore/internal/benchmark:kvstore_benchmark

LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libprofiler.so.0.5.1 \
CPUPROFILE=http.prof \
./bazel-bin/tensorstore/internal/benchmark/kvstore_benchmark \
   --kvstore_spec='{"driver": "gcs", "bucket":"<MY_GCS_BUCKET>", "path":"default"}' \
   --chunk_size=67108864 \
   --total_bytes=67108864000 \
   --repeat_reads=10 \
   --repeat_writes=0
   
google-pprof --svg ./bazel-bin/tensorstore/internal/benchmark/kvstore_benchmark http.prof > prof.svg
```

### I expected the following

I didn't expect 31% of the time to live in `Curl_bufq_len`

In fact, if I remove the  applicable clause in `http2.c:on_stream_frame`, then the transfer rate nearly doubles.

### curl/libcurl version

curl 8.4.0
curl 8.6.0

### operating system

GCE VM.

5.10.0-26-cloud-amd64 #1 SMP Debian 5.10.197-1 (2023-09-29) x86_64 GNU/Linux

**Comment by laramiel** [comment]
The code paths have changed for 8.7.0; this may be obsolete.

**Comment by bagder** [comment]
Probably even more in current master. Would be interesting to know if there is any problems there... @icing have you run your benchmarks recently?

**Comment by icing** [comment]
The observation fits into the implementation restrictions that I talked about in  https://github.com/icing/blog/blob/main/curl-h2-perf-evolution.md 

tl;dr

There is no longer a `bufq` holding the response data, it is written directly to the application callbacks now.

**Comment by laramiel** [comment]
Verified that the throughput for this case is improved by ~60% along with one reduced copy & removal of Curl_bufq_len in the hot path post 8.7.0.

Thanks.
